src/Microsoft.AspNetCore.Server.Kestrel/Filter/StreamSocketOutput.cs

```diff
@@ -62,17 +80,39 @@ public MemoryPoolIterator ProducingStart()
 
         public void ProducingComplete(MemoryPoolIterator end)
         {
-            var block = _producingBlock;
-            while (block != end.Block)
+            lock (_writeLock)
             {
-                _outputStream.Write(block.Data.Array, block.Data.Offset, block.Data.Count);
+                ProducingCompleteAsync(end).GetAwaiter().GetResult();
```

Comment:
The main difference is that the real ProducingComplete would never wait for the write behind buffer to flush. This might by waiting for the Task returned by SocketOutput.WriteAsync. It's a difference in behavior that we probably don't account for in the calling code.